### PR TITLE
Support ansible 2.7 style include_tasks

### LINKED ISSUE #507: ansible-lint errors when using 2.7+ include_tasks apply argument
# Issue Type
- Bug report

# Ansible and Ansible Lint details

    $ ansible --version
    ansible 2.7.10
      config file = /etc/ansible/ansible.cfg
      configured module search path = ['/home/ansible/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
      ansible python module location = /usr/local/lib/python3.7/dist-packages/ansible
      executable location = /usr/local/bin/ansible
      python version = 3.7.3rc1 (default, Mar 13 2019, 11:01:15) [GCC 8.3.0]

    $ ansible-lint --version
    ansible-lint 4.1.0

- installation method: `pip3 install ansible ansible-lint` inside debian:buster container from docker hub

# Desired Behaviour

Shouldn't fail on valid syntax for 2.7+

- https://docs.ansible.com/ansible/latest/modules/include_tasks_module.html#parameters

# Actual Behaviour (Bug report only)

Example Files

    $ more *
    ::::::::::::::
    playbook.yml
    ::::::::::::::
    - hosts: localhost
      gather_facts: no
      tasks:
      - name: simple include task with apply
        include_tasks:
          file: debug.yml
          apply:
            tags:
              - debug
    ::::::::::::::
    debug.yml
    ::::::::::::::
    - debug:

    ::::::::::::::
    roles/example/main.yml
    ::::::::::::::
    - include_tasks:
        file: debug.yml
        apply:
          tags:
          - debug
    ::::::::::::::
    roles/example/debug.yml
    ::::::::::::::
    - debug:

Errors from ansible-lint 

    $ ansible-lint roles/example/
    WARNING: Couldn't open /home/ansible/project/roles/example/tasks/{'file': - No such file or directory

    $ ansible-lint playbook.yml 
    Traceback (most recent call last):
      File "/usr/local/bin/ansible-lint", line 10, in <module>
        sys.exit(main())
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/__main__.py", line 187, in main
        matches.extend(runner.run())
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/__init__.py", line 264, in run
        for child in ansiblelint.utils.find_children(arg, self.playbook_dir):
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 171, in find_children
        for child in play_children(basedir, item, playbook[1], playbook_dir):
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 223, in play_children
        return delegate_map[k](basedir, k, v, parent_type)
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 243, in _taskshandlers_children
        append_children(th['include_tasks'], basedir, k, parent_type, results)
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 273, in append_children
        'path': path_dwim(basedir, taskhandler),
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 72, in path_dwim
        return dl.path_dwim(given)
      File "/usr/local/lib/python3.7/dist-packages/ansible/parsing/dataloader.py", line 181, in path_dwim
        given = unquote(given)
      File "/usr/local/lib/python3.7/dist-packages/ansible/parsing/quoting.py", line 29, in unquote
        if is_quoted(data):
      File "/usr/local/lib/python3.7/dist-packages/ansible/parsing/quoting.py", line 24, in is_quoted
        return len(data) > 1 and data[0] == data[-1] and data[0] in ('"', "'") and data[-2] != '\\'
    KeyError: 0

Yet these are a valid playbook and role

    $ ansible-playbook -i localhost, playbook.yml 

    PLAY [localhost] ******************************************************************************************************

    TASK [simple include task with apply] *********************************************************************************
    included: /home/ansible/project/debug.yml for localhost

    TASK [debug] **********************************************************************************************************
    ok: [localhost] => {
        "msg": "Hello world!"
    }

    PLAY RECAP ************************************************************************************************************
    localhost  

    $ ansible localhost -i localhost, -m include_role -a name=roles/example
    localhost | SUCCESS => {
        "changed": false,
        "include_variables": {
            "name": "roles/example"
        }
    }
    localhost | SUCCESS => {
        "changed": false,
        "include": "debug.yml",
        "include_variables": {
            "apply": {
                "tags": [
                    "debug"
                ]
            }
        }
    }
    localhost | SUCCESS => {
        "msg": "Hello world!"
    }

### PR DESCRIPTION
**Pull request by gdoucet**
Fixes #507

Signed-off-by: Geoffroy Doucet <geoffroy.doucet@kloodz.com>

### CODE REVIEW

**Comment by gdoucet** [comment]
Can someone take a look at this pull request?

**Comment by ssbarnea** [comment]
@webknjaz @awcrosby please.

**Comment by gdoucet** [comment]
Any news on this pull request?

**Comment by ssbarnea** [comment]
@gdoucet Please post link to irc on #ansible-galaxy and #ansible-devel channels, it should help.

File: lib/ansiblelint/utils.py [hunk 1]
```diff
@@ -241,7 +241,10 @@ def _taskshandlers_children(basedir, k, v, parent_type):
         if 'include' in th:
             append_children(th['include'], basedir, k, parent_type, results)
         elif 'include_tasks' in th:
-            append_children(th['include_tasks'], basedir, k, parent_type, results)
+            if 'file' in th['include_tasks']:
```
Comment (webknjaz) [inline lib/ansiblelint/utils.py]: To much duplication.
```suggestion
            append_children(
                th['include_tasks'].get('file', th['include_tasks']),
                basedir,
                k,
                parent_type,
                results
            )
```

File: test/TestTaskIncludes.py [hunk 2]
```diff
@@ -56,3 +56,10 @@ def test_include_tasks_in_role(self):
         runner = Runner(self.rules, filename, [], [], [])
         runner.run()
         self.assertEqual(len(runner.playbooks), 4)
+
+    @unittest.skipIf(parse_version(ansible.__version__) < parse_version('2.7'), "not supported with ansible < 2.7")
+    def test_import_tasks_2_7_style(self):
+        filename = 'test/taskincludes_2_7_style.yml'
+        runner = Runner(self.rules, filename, [], [], [])
+        runner.run()
+        self.assertEqual(len(runner.playbooks), 2)
```
Comment (webknjaz) [inline test/TestTaskIncludes.py]: Now that we use pytest runner, it's recommended to use `assert`
```suggestion
        assert len(runner.playbooks) == 2
```

File: test/taskincludes_2_7_style.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+- hosts: webservers
+  vars:
+    varset: varset
+  tasks:
+    - include_tasks:
+        file: "{{ varset }}.yml"
+    - include_tasks:
+        file: "directory with spaces/main.yml"
```
Comment (webknjaz) [inline test/taskincludes_2_7_style.yml]: You only test the code branch that always has `file`. Please also cover the case w/o it.

File: test/taskincludes_2_7_style.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+- hosts: webservers
+  vars:
+    varset: varset
+  tasks:
+    - include_tasks:
```
Comment (webknjaz) [inline test/taskincludes_2_7_style.yml]: Also test negative cases with invalid values
```suggestion
    - include_tasks:
    - include_tasks:
```

File: test/TestTaskIncludes.py (hunk 2 above)
Comment (ssbarnea) [reply test/TestTaskIncludes.py]: I do remember that assert was not encouraged for several reasons but I forgot why. I think one may be the way pytest displays a potential assert failure?

On the other hand I do like the clearness of a simple assert statement.

Comment (webknjaz) [reply test/TestTaskIncludes.py]: `assert` is not encouraged in runtime because it's stripped off in the optimized mode (`python -O`). Under the hood, it's implemented as
```python
if __debug__:
    raise AssertError if needed
```
And `__debug__` is `False` in the optimized mode.

This was valid for nose/unittest runners but `pytest` specifically relies on `assert`s and it also rewrites the bytecode for them.
Displaying extra assertion comparison is another feature with arises from this.

File: lib/ansiblelint/utils.py [hunk 5]
```diff
@@ -241,7 +241,19 @@ def _taskshandlers_children(basedir, k, v, parent_type):
...
+                    th['include_tasks']
+                ),
+                basedir,
+                k,
+                parent_type,
+                results
+            )
+                append_children(th['include_tasks']['file'], basedir, k, parent_type, results)
+            else:
+                append_children(th['include_tasks'], basedir, k, parent_type, results)
```
Comment (Yajo) [inline lib/ansiblelint/utils.py]: There's something wrong with indentation in these lines.

**Comment by patrickjahns** [comment]
@gdoucet - are you still working on the requested changes? If not would you fine with me taking over this PR ?

**Comment by webknjaz** [comment]
@patrickjahns I think it's safe to proceed with your own PR

File: lib/ansiblelint/utils.py (hunk 5 above)
Comment (webknjaz) [reply lib/ansiblelint/utils.py]: That suggested change was supposed to hit more lines than one but at the time GH didn't support multiline comments.

Comment (webknjaz) [inline lib/ansiblelint/utils.py]:
```suggestion
```

File: lib/ansiblelint/utils.py [hunk 6]
```diff
@@ -241,7 +241,16 @@ def _taskshandlers_children(basedir, k, v, parent_type):
...
-            append_children(th['include_tasks'], basedir, k, parent_type, results)
+            append_children(
+                th['include_tasks'].get(
+                    'file',
+                    th['include_tasks']
+                ),
+                basedir,
+                k,
+                parent_type,
+                results
```
Comment (webknjaz) [inline lib/ansiblelint/utils.py]:
```suggestion
                    th['include_tasks'],
                ),
                basedir,
                k,
                parent_type,
                results,
```

File: test/TestTaskIncludes.py [hunk 7]
```diff
@@ -56,3 +56,10 @@ def test_include_tasks_in_role(self):
         runner = Runner(self.rules, filename, [], [], [])
         runner.run()
         self.assertEqual(len(runner.playbooks), 4)
+
+    @unittest.skipIf(parse_version(ansible.__version__) < parse_version('2.7'), "not supported with ansible < 2.7")
```
Comment (webknjaz) [inline test/TestTaskIncludes.py]: Since we're moving towards pytest-based tests, we should avoid using `unittest` module.
Here's an example of a test module using only `pytest`: https://github.com/ansible/ansible-lint/blob/41d597d0583967fc21c0f1b5a1281b4205b14d37/test/TestIncludeMissingFileRule.py.

Comment (ssbarnea) [reply test/TestTaskIncludes.py]: @webknjaz Thanks for helping with that. I still see few conflicts.

Comment (webknjaz) [reply test/TestTaskIncludes.py]: Yeah, I was doing all the coding in the browser. This needs to be rebased manually. But also we need other comments to be addressed. So maybe if @patrickjahns takes over, he'll do this in his PR. I only committed this change to avoid going back and forth with pytest coding style and people being unfamiliar w/ it. Other things like almost missing test coverage would still have to be fixed by somebody.

Comment (ssbarnea) [reply test/TestTaskIncludes.py]: Editing PRs in browser, that is a brave task for anything more than a one liner.

**Comment by ssbarnea** [comment]
@gdoucet Can you please fix conflicts on this? As we dropped support for 2.7, the change should be even less complex.

**Comment by ssbarnea** [comment]
I am closing this because it was not fixed in a very long time. If @gdoucet can fix it, I would be more than happy to review it.
